cpp/include/cudf/utilities/traits.hpp

```diff
@@ -0,0 +1,182 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/types.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <type_traits>
+
+namespace cudf {
+
+template <typename...>
+using void_t = void;
+
+template <typename L, typename R, typename = void>
+struct is_relationally_comparable_impl : std::false_type {};
+
+template <typename L, typename R>
+using less_comparable = decltype(std::declval<L>() < std::declval<R>());
+
+template <typename L, typename R>
+using greater_comparable = decltype(std::declval<L>() > std::declval<R>());
+
+template <typename L, typename R>
+struct is_relationally_comparable_impl<
+    L, R, void_t<less_comparable<L, R>, greater_comparable<L, R> > >
+    : std::true_type {};
+
+/**---------------------------------------------------------------------------*
+ * @brief Indicates if objects of types `L` and `R` can be relationally
+ *compared.
+ *
+ * Given two objects `L l`, and `R r`, returns true if `l < r` and `l > r` are
+ * well-formed expressions.
+ *
+ * @tparam L Type of the first object
+ * @tparam R Type of the second object
+ * @return true Objects of types `L` and `R` can be relationally be compared
+ * @return false Objects of types `L` and `R` cannot be compared
+ *---------------------------------------------------------------------------**/
+template <typename L, typename R>
+constexpr inline bool is_relationally_comparable() {
+  return is_relationally_comparable_impl<L, R>::value;
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief Indicates if the type `T` is a numeric type.
+ *
+ * @tparam T  The type to verify
+ * @return true `T` is numeric
+ * @return false  `T` is not numeric
+ *---------------------------------------------------------------------------**/
+template <typename T>
+constexpr inline bool is_numeric() {
+  return std::is_integral<T>::value or std::is_floating_point<T>::value;
+}
+
+/**---------------------------------------------------------------------------*
+ * @brief Indicates if `type` is a numeric `data_type`.
+ *
+ * "Numeric" types are fundamental integral/floating point types such as `INT*`
+ * or `FLOAT*`. Types that wrap a numeric type are not considered numeric, e.g.,
+ *`TIMESTAMP` or `DATE32`.
+ *
+ * @param type The `data_type` to verify
+ * @return true `type` is numeric
+ * @return false `type` is not numeric
+ *---------------------------------------------------------------------------**/
+constexpr inline bool is_numeric(data_type type) {
+  return cudf::exp::type_dispatcher(
+      type, [](auto dummy) { return is_numeric<decltype(dummy)>(); }, 0);
```

Comment:
Will passing 0 work with wrapper types which need to be explicitly constructed?

Reply:
Yes, it should. However, I do not recommend using this pattern for the `type_dispatcher` in general. I did it here to save on cluttering this file with a bunch of function object definitions.